File: cpp/src/msgpack/rpc/future.cc
//
// msgpack::rpc::future - MessagePack-RPC for C++
//
// Copyright (C) 2010 FURUHASHI Sadayuki
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
//
#include "future_impl.h"
#include <cclog/cclog.h>

namespace msgpack {
namespace rpc {


void future_impl::wait()
{
	mp::pthread_scoped_lock lk(m_mutex);
	while(m_session) {
		m_cond.wait(m_mutex);
	}
}

void future_impl::recv()
{
	while(m_session) {
		m_loop->run_once();
	}
}

void future_impl::join()
{
	if(m_loop->is_running()) {
		wait();
	} else {
		recv();
	}
}


object future_impl::get_impl()
{
	join();
	if(!m_error.is_nil()) {
		// FIXME throw msgpack::rpc::remote_error
		throw std::runtime_error("remote error");
	}
	return m_result;
}

object future::get_impl()
{
	if(!m_pimpl) {
		// FIXME
		throw std::runtime_error("null future reference");
	}
	return m_pimpl->get_impl();
}


static void callback_real(
		callback_t callback, future f)
try {
	callback(f);
} catch (std::exception& e) {
	LOG_ERROR("response callback error: ",e.what());
} catch (...) {
	LOG_ERROR("response callback error: unknown error");
}

void future_impl::attach_callback(callback_t func)
{
	mp::pthread_scoped_lock lk(m_mutex);
	m_callback = func;
	if(!m_session) {
		lk.unlock();
		m_loop->submit(&callback_real, m_callback,
				future(shared_from_this()));
	}
}


void future_impl::set_result(object result, object error, auto_zone z)
{
	mp::pthread_scoped_lock lk(m_mutex);
	m_result = result;
	m_error = error;
	m_z = z;
	m_session.reset();

	m_cond.broadcast();

	if(m_callback) {
		lk.unlock();
		// FIXME submit?
		callback_real(m_callback, future(shared_from_this()));
	}
}


future& future::join()
{
	m_pimpl->join();
	return *this;
}

future& future::wait()
{
	m_pimpl->wait();
	return *this;
}

future& future::recv()
{
	m_pimpl->recv();
	return *this;
}

future& future::attach_callback(
		mp::function<void (future)> func)
{
	m_pimpl->attach_callback(func);
	return *this;
}


}  // namespace rpc
}  // namespace msgpack

File: ruby/lib/msgpack/rpc/future.rb
#
# MessagePack-RPC for Ruby
#
# Copyright (C) 2010-2011 FURUHASHI Sadayuki
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#
module MessagePack
module RPC


# Future describes result of remote procedure call that is initially not known,
# because it is not yet received.
# You can wait and get the result with get method.
class Future
	def initialize(session, loop, callback = nil)  #:nodoc:
		@timeout = session.timeout
		@loop = loop
		@callback_handler = callback
		@error_handler = nil
		@result_handler = nil
		@set = false
		@error = nil
		@result = nil
	end
	attr_reader :loop
	attr_accessor :result, :error

	# Wait for receiving result of remote procedure call and returns its result.
	# If the remote method raises error, then this method raises RemoteError.
	# If the remote procedure call failed with timeout, this method raises TimeoutError.
	# Otherwise this method returns the result of remote method.
	def get
		join
		if error.nil?
			if @result_handler
				return @result_handler.call(@result)
			else
				return @result
			end
		end
		if @result.nil?
			# compatible error
			raise RuntimeError.new(@error)
		end
		if @error_handler
			@error_handler.call(@error, @result)
		end
		raise RPCError.create(@error, @result)
	end

	# Wait for receiving result of remote procedure call.
	# This method returns self.
	# If a callback method is attached, it will be called.
	def join
		until @set
			@loop.run_once
		end
		self
	end

	# call-seq:
	#   attach_callback {|future|  }
	#
	# Attaches a callback method that is called when the result of remote method is received.
	def attach_callback(proc = nil, &block)
		@callback_handler = proc || block
	end

	# For IDL
	def attach_error_handler(proc = nil, &block)  #:nodoc:
		@error_handler = proc || block
	end

	# For IDL
	def attach_result_handler(proc = nil, &block)  #:nodoc:
		@result_handler = proc || block
	end

	def set_result(err, res)  #:nodoc:
		@error  = err
		@result = res
		@set = true
		if @callback_handler
			if @callback_handler.arity == 2
				# FIXME backward compatibility
				@callback_handler.call(error, result)
			else
				@callback_handler.call(self)
			end
		end
		self
	rescue
		self
	end

	def step_timeout  #:nodoc:
		if @timeout < 1
			true
		else
			@timeout -= 1
			false
		end
	end
end


end
end
